Build failure with MinGW-w64 gcc 5.2.0

## Updated description

We need to document that the standard MSYS2 package for gcc **will not work** when building Rust, you have to install a custom toolchain with a known good version and the win32 threading model, not the pthread threading model.

## Original report

Looks related to https://github.com/rust-lang/rust/pull/29177, perhaps the `gcc_eh` library picked up new dependencies in 5.2.0?

cc @vadimcn 

```
error: linking with `gcc` failed: exit code: 1
note: "gcc" "-Wl,--enable-long-section-names" "-fno-use-linker-plugin" "-Wl,--nxcompat" "-nostdlib" "-Wl,--large-address-aware" "C:\\msys64\\home\\ale\\rust2\\i686-pc-windows-gnu\\stage1\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\dllcrt2.o" "C:\\msys64\\home\\ale\\rust2\\i686-pc-windows-gnu\\stage1\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\rsbegin.o" "-L" "C:\\msys64\\home\\ale\\rust2\\i686-pc-windows-gnu\\stage1\\lib\\rustlib\\i686-pc-windows-gnu\\lib" "i686-pc-windows-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib\\std-71b07a99.0.o" "-o" "i686-pc-windows-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib\\std-71b07a99.dll" "i686-pc-windows-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib\\std-71b07a99.metadata.o" "-nodefaultlibs" "-L" "i686-pc-windows-gnu/rt" "-L" "C:\\msys64\\home\\ale\\rust2\\i686-pc-windows-gnu\\llvm\\Release/lib" "-L" "C:\\msys64\\home\\ale\\rust2\\i686-pc-windows-gnu\\stage1\\lib\\rustlib\\i686-pc-windows-gnu\\lib" "-L" "C:\\msys64\\home\\ale\\rust2\\.rust\\lib\\i686-pc-windows-gnu" "-L" "C:\\msys64\\home\\ale\\rust2\\lib\\i686-pc-windows-gnu" "-Wl,-Bstatic" "-Wl,--whole-archive" "-l" "backtrace" "-Wl,--no-whole-archive" "-Wl,-Bdynamic" "-l" "gcc_eh" "-l" "gcc_eh" "-l" "ws2_32" "-l" "userenv" "-l" "shell32" "-l" "advapi32" "-Wl,--whole-archive" "C:\\msys64\\tmp\\rustc.hZLinbW19bbD\\libcollections-71b07a99.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\msys64\\tmp\\rustc.hZLinbW19bbD\\liballoc-71b07a99.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\msys64\\tmp\\rustc.hZLinbW19bbD\\librustc_unicode-71b07a99.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\msys64\\tmp\\rustc.hZLinbW19bbD\\liballoc_jemalloc-71b07a99.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\msys64\\tmp\\rustc.hZLinbW19bbD\\liblibc-71b07a99.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\msys64\\tmp\\rustc.hZLinbW19bbD\\librand-71b07a99.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\msys64\\tmp\\rustc.hZLinbW19bbD\\libcore-71b07a99.rlib" "-Wl,--no-whole-archive" "-shared" "-l" "compiler-rt" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-luser32" "-lkernel32" "C:\\msys64\\home\\ale\\rust2\\i686-pc-windows-gnu\\stage1\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\rsend.o"
note: C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2.o):(.text+0x1a4b): undefined reference to `pthread_once'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x1a3): undefined reference to `pthread_mutex_init'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0xf99): undefined reference to `pthread_once'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0xfa5): undefined reference to `pthread_mutex_lock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x10b4): undefined reference to `pthread_once'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x10c0): undefined reference to `pthread_mutex_lock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x118b): undefined reference to `pthread_once'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x1197): undefined reference to `pthread_mutex_lock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x11f9): undefined reference to `pthread_mutex_unlock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x1228): undefined reference to `pthread_mutex_unlock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x12af): undefined reference to `pthread_once'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x12bb): undefined reference to `pthread_mutex_lock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x12f7): undefined reference to `pthread_mutex_unlock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x13b6): undefined reference to `pthread_mutex_unlock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0xfc4): undefined reference to `pthread_mutex_unlock'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/5.2.0/libgcc_eh.a(unwind-dw2-fde.o):(.text+0x10df): undefined reference to `pthread_mutex_unlock'
collect2.exe: error: ld returned 1 exit status

error: aborting due to previous error

This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
/home/ale/rust2/mk/target.mk:231: recipe for target 'i686-pc-windows-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib/stamp.std' failed
make: *** [i686-pc-windows-gnu/stage1/lib/rustlib/i686-pc-windows-gnu/lib/stamp.std] Error 3
```

--- a/README.md
+++ b/README.md
@@ -53,6 +53,16 @@ Read ["Installing Rust"] from [The Book].
 
 ### Building on Windows
 
+There are two prominent ABIs in use on Windows: the native (MSVC) ABI used by
+Visual Studio, and the GNU ABI used by the GCC toolchain. Which version of Rust
+you need depends largely on what C/C++ libraries you want to interoperate with:
+for interop with software produced by Visual Studio use the MSVC build of Rust;
+for interop with GNU software built using the MinGW/MSYS2 toolchain use the GNU
+build.
+
+
+#### MinGW
+
 [MSYS2](http://msys2.github.io/) can be used to easily build Rust on Windows:
 
 1. Grab the latest MSYS2 installer and go through the installer.
@@ -63,12 +73,15 @@ Read ["Installing Rust"] from [The Book].
    ```sh
    # Update package mirrors (may be needed if you have a fresh install of MSYS2)
    $ pacman -Sy pacman-mirrors
+   ```
 
-   # Choose one based on platform: 
-   # *** see the note below ***
-   $ pacman -S mingw-w64-i686-toolchain
-   $ pacman -S mingw-w64-x86_64-toolchain
+Download [MinGW from
+here](http://mingw-w64.org/doku.php/download/mingw-builds), and choose the
+`threads=win32,exceptions=dwarf/seh` flavor when installing. After installing,
+add its `bin` directory to your `PATH`. This is due to #28260, in the future,
+installing from pacman should be just fine.
 
+   ```
    # Make git available in MSYS2 (if not already available on path)
    $ pacman -S git
 
@@ -84,16 +97,19 @@ Read ["Installing Rust"] from [The Book].
    $ ./configure
    $ make && make install
    ```
-> ***Note:*** gcc versions >= 5 currently have issues building LLVM on Windows
-> resulting in a segmentation fault when building Rust. In order to avoid this
-> it may be necessary to obtain an earlier version of gcc such as 4.9.x.  
-> Msys's `pacman` will install the latest version, so for the time being it is
-> recommended to skip gcc toolchain installation step above and use [Mingw-Builds]
-> project's installer instead.  Be sure to add gcc `bin` directory to the path
-> before running `configure`.  
-> For more information on this see issue #28260.
 
-[Mingw-Builds]: http://sourceforge.net/projects/mingw-w64/
+#### MSVC
+
+MSVC builds of Rust additionally require an installation of Visual Studio 2013
+(or later) so `rustc` can use its linker. Make sure to check the “C++ tools”
+option. In addition, `cmake` needs to be installed to build LLVM.
+
+With these dependencies installed, the build takes two steps:
+
+```sh
+$ ./configure
+$ make && make install
+```
 
 ## Building Documentation
 

